### PR TITLE
Removing floats from the ML library and tests

### PR DESCRIPTION
**Pull request by evsmithx**
Removing floats from the ml library. This PR depends on #2137 and #2127 (which have both now gone in). All the ml tests have been defloated apart from the ops ones.

### CODE REVIEW

File: libs/ml/include/ml/optimisation/rmsprop_optimiser.hpp [hunk 1]
```diff
@@ -42,16 +42,16 @@ class RMSPropOptimiser : public Optimiser<T>
...
+                   DataType const &epsilon       = fetch::math::Type<DataType>("1e-8"));
 
   RMSPropOptimiser(std::shared_ptr<Graph<T>>       graph,
                    std::vector<std::string> const &input_node_names,
                    std::string const &label_node_name, std::string const &output_node_name,
                    fetch::ml::optimisers::LearningRateParam<DataType> const &learning_rate_param,
-                   DataType const &decay_rate = DataType{0.9f},
-                   DataType const &epsilon    = DataType{1e-8f});
+                   DataType const &decay_rate = fetch::math::Type<DataType>("0.9"),
+                   DataType const &epsilon    = fetch::math::Type<DataType>("1e-8"));
```
Comment (vhavrylov-ncube) [inline libs/ml/include/ml/optimisation/rmsprop_optimiser.hpp]: Scientific notation

File: libs/ml/include/ml/optimisation/rmsprop_optimiser.hpp [hunk 2]
```diff
@@ -42,16 +42,16 @@ class RMSPropOptimiser : public Optimiser<T>
   RMSPropOptimiser(std::shared_ptr<Graph<T>>       graph,
                    std::vector<std::string> const &input_node_names,
                    std::string const &label_node_name, std::string const &output_node_name,
-                   DataType const &learning_rate = DataType{0.001f},
-                   DataType const &decay_rate    = DataType{0.9f},
-                   DataType const &epsilon       = DataType{1e-8f});
+                   DataType const &learning_rate = fetch::math::Type<DataType>("0.001"),
+                   DataType const &decay_rate    = fetch::math::Type<DataType>("0.9"),
+                   DataType const &epsilon       = fetch::math::Type<DataType>("1e-8"));
```
Comment (vhavrylov-ncube) [inline libs/ml/include/ml/optimisation/rmsprop_optimiser.hpp]: Scientific notation

File: libs/ml/tests/unit/graph/graph.cpp [hunk 3]
```diff
@@ -369,25 +369,25 @@ TYPED_TEST(GraphTest, variable_freezing_subgraph)
...
-                                       static_cast<DataType>(0)));
-  ASSERT_FALSE(weights_1.at(2).AllClose(weights_2.at(2), static_cast<DataType>(0),
-                                        static_cast<DataType>(0)));
-  ASSERT_FALSE(weights_1.at(3).AllClose(weights_2.at(3), static_cast<DataType>(0),
-                                        static_cast<DataType>(0)));
-  ASSERT_TRUE(weights_1.at(4).AllClose(weights_2.at(4), static_cast<DataType>(0),
-                                       static_cast<DataType>(0)));
-  ASSERT_TRUE(weights_1.at(5).AllClose(weights_2.at(5), static_cast<DataType>(0),
-                                       static_cast<DataType>(0)));
+  ASSERT_TRUE(weights_1.at(0).AllClose(weights_2.at(0), fetch::math::Type<DataType>("0"),
```
Comment (vhavrylov-ncube) [inline libs/ml/tests/unit/graph/graph.cpp]: (optional) DataType{0} could be a bit more clear here, especially as there are only integers in this code block

File: libs/ml/tests/unit/graph/graph.cpp [hunk 4]
```diff
@@ -397,24 +397,24 @@ TYPED_TEST(GraphTest, variable_freezing_subgraph)
...
-                                        static_cast<DataType>(0)));
-  ASSERT_FALSE(weights_2.at(2).AllClose(weights_3.at(2), static_cast<DataType>(0),
-                                        static_cast<DataType>(0)));
-  ASSERT_FALSE(weights_2.at(3).AllClose(weights_3.at(3), static_cast<DataType>(0),
-                                        static_cast<DataType>(0)));
-  ASSERT_FALSE(weights_2.at(4).AllClose(weights_3.at(4), static_cast<DataType>(0),
-                                        static_cast<DataType>(0)));
-  ASSERT_FALSE(weights_2.at(5).AllClose(weights_3.at(5), static_cast<DataType>(0),
-                                        static_cast<DataType>(0)));
+  ASSERT_FALSE(weights_2.at(0).AllClose(weights_3.at(0), fetch::math::Type<DataType>("0"),
```
Comment (vhavrylov-ncube) [inline libs/ml/tests/unit/graph/graph.cpp]: (optional) DataType{0}

**Review by vhavrylov-ncube — DISMISSED** [review]

File: libs/ml/examples/commodity_prediction/main.cpp [hunk 5]
```diff
@@ -462,11 +464,11 @@ int main(int argc, char **argv)
...
     loader.AddData({data}, label);
 
-    DataType    distance         = 0;
-    DataType    distance_counter = 0;
+    DataType    distance{0};
+    DataType    distance_counter{0};
     std::string our_y_output = filename_root + "y_pred_test_fetch_" + std::to_string(EPOCHS) + "_" +
                                std::to_string(static_cast<int>(use_random)) + "_" +
-                               std::to_string(LEARNING_RATE) + ".csv";
+                               std::to_string(static_cast<double>(LEARNING_RATE)) + ".csv";
```
Comment (baykaner) [inline libs/ml/examples/commodity_prediction/main.cpp:471]: why do you need to cast to double and then to string? why not directly to string?

File: libs/ml/include/ml/model/model_config.hpp [hunk 6]
```diff
@@ -31,7 +31,7 @@ struct ModelConfig
   bool     early_stopping = false;
   bool     test           = false;
   SizeType patience       = 10;
-  DataType min_delta      = DataType(0.0);
+  DataType min_delta      = fetch::math::Type<DataType>("0.0");
```
Comment (baykaner) [inline libs/ml/include/ml/model/model_config.hpp]: or even DataType{0} should be fine

File: libs/ml/include/ml/ops/activations/gelu.hpp [hunk 7]
```diff
@@ -90,10 +90,8 @@ class Gelu : public Ops<T>
     TensorType        intermediate1({input.shape()}), intermediate2({input.shape()}),
         intermediate3({input.shape()});
 
-    DataType one{static_cast<DataType>(1)}, two{static_cast<DataType>(2)},
-        neg_two{static_cast<DataType>(-2)}, three{static_cast<DataType>(3)},
-        half{static_cast<DataType>(0.5)}, a{static_cast<DataType>(0.797885)},
-        b{static_cast<DataType>(0.035677)};
+    DataType one{1}, two{2}, neg_two{-2}, three{3}, half{fetch::math::Type<DataType>("0.5")},
```
Comment (baykaner) [inline libs/ml/include/ml/ops/activations/gelu.hpp]: style: one declaration per line

File: libs/ml/include/ml/ops/trainable.hpp [hunk 8]
```diff
@@ -59,7 +59,8 @@ class Trainable
   virtual void ApplySparseGradient(TensorType const &grad, SizeSet &update_rows)           = 0;
   virtual void ApplyRegularisation()                                                       = 0;
 
-  void SetRegularisation(RegPtrType regulariser, DataType regularisation_rate = DataType{0.0})
+  void SetRegularisation(RegPtrType regulariser,
+                         DataType   regularisation_rate = fetch::math::Type<DataType>("0.0"))
```
Comment (baykaner) [inline libs/ml/include/ml/ops/trainable.hpp]: DataType{0} should be fine?

File: libs/ml/include/ml/ops/trainable.hpp [hunk 9]
```diff
@@ -81,7 +82,7 @@ class Trainable
 
 protected:
   RegPtrType regulariser_;
-  DataType   regularisation_rate_ = static_cast<DataType>(0);
+  DataType   regularisation_rate_ = fetch::math::Type<DataType>("0");
```
Comment (baykaner) [inline libs/ml/include/ml/ops/trainable.hpp]: DataType   regularisation_rate_{0};

File: libs/ml/tests/unit/graph/graph.cpp [hunk 10]
```diff
@@ -147,13 +147,13 @@ TYPED_TEST(GraphTest, multi_nodes_have_same_name)
   std::string input = g.template AddNode<fetch::ml::ops::PlaceHolder<TensorType>>("Input", {});
   std::string fc_1  = g.template AddNode<fetch::ml::layers::FullyConnected<TensorType>>(
       "FC1", {input}, 10u, 10u, fetch::ml::details::ActivationType::NOTHING,
-      fetch::ml::RegularisationType::NONE, static_cast<DataType>(0));
+      fetch::ml::RegularisationType::NONE, fetch::math::Type<DataType>("0"));
```
Comment (baykaner) [inline libs/ml/tests/unit/graph/graph.cpp]: so... there wasn't really anything wrong with what was already there: static_cast<DataType>(0) is an integer constructor and has no indeterminacy risk. If you really want to replace it you could swap it out for DataType{0}, but there's no reason to call the string constructor here

File: libs/ml/tests/unit/graph/graph.cpp (hunk 4 above)
Comment (baykaner) [reply libs/ml/tests/unit/graph/graph.cpp]: I would say that's not optional; static_cast<DataType>(0) is better than the proposed change; DataType{0} is also fine

**Review by baykaner — DISMISSED** [review]

File: libs/ml/examples/commodity_prediction/main.cpp (hunk 5 above)
Comment (evsmithx) [reply libs/ml/examples/commodity_prediction/main.cpp:471]: I don't think that I can cast a fixed point to a string

File: libs/ml/include/ml/ops/activations/gelu.hpp (hunk 7 above)
Comment (evsmithx) [reply libs/ml/include/ml/ops/activations/gelu.hpp]: Fixed in math PR

Comment (evsmithx) [reply libs/ml/include/ml/ops/activations/gelu.hpp]: Ignore the above!

File: libs/ml/examples/bert_for_finetune/main.cpp [hunk 11]
```diff
@@ -67,7 +67,7 @@ int main(int ac, char **av)
   SizeType batch_size = 4;
   SizeType epochs     = 20;
   SizeType layer_no   = 12;
-  auto     lr         = static_cast<DataType>(1e-5);
+  auto     lr         = static_cast<DataType>(0.00001);
```
Comment (vhavrylov-ncube) [inline libs/ml/examples/bert_for_finetune/main.cpp]: fetch::math::Type<DataType>("0.00001")

Comment (evsmithx) [reply libs/ml/examples/bert_for_finetune/main.cpp]: Good catch!

File: libs/ml/tests/unit/layers/scaled_dot_product_attention.cpp [hunk 12]
```diff
@@ -198,13 +198,13 @@ TYPED_TEST(ScaledDotProductAttention,
   g.SetInput(mask, mask_data);
 
   TypeParam gt = TypeParam::FromString(
-      "1.8496745531,  1.9944926680,  1.5288354812,  0.1000000000, 0.1000000000,  0.1000000000; "
-      "1.1503254469,  1.0055073320,  1.4711645188, -0.2000000000, -0.2000000000, -0.2000000000; "
-      "3.6993491062,  3.9889853359,  3.0576709623,  1.0000000000, 1.0000000000,  1.0000000000");
+      "+1.848551915, +1.994484604, +1.323288694, +2.134702830, +1.607807817, "
+      "+3.647750064;+1.149618159, +1.005503525, +1.237182403, -0.486945271, -0.412639564, "
+      "-0.700323727;+3.696272045, +3.988963812, +2.446791525, -0.017351415, +0.246096091, "
+      "-0.773875033;");
```
Comment (MissingNO57) [inline libs/ml/tests/unit/layers/scaled_dot_product_attention.cpp]: This seems weird because mask should block gradient, so these values  0.1000000000, 0.1000000000,  0.1000000000; etc. should be the same.

File: libs/ml/tests/unit/layers/scaled_dot_product_attention.cpp [hunk 13]
```diff
@@ -235,19 +235,22 @@ TYPED_TEST(ScaledDotProductAttention,
...
+      "+0.000000000;");
   gt_key_grad.Reshape({3, 3, 2});
   TypeParam gt_value_grad = TypeParam::FromString(
-      "0.1558327790,  1.8441672210,  0.0000000000, -1.5000000000, 0.0000000000,  0.0000000000; "
-      "0.1668474430,  3.8331525570,  0.0000000000,  4.0000000000, 0.0000000000,  0.0000000000; "
-      "0.1640937770,  3.3359062230,  0.0000000000,  0.0000000000, 0.0000000000,  0.0000000000");
+      "+0.155707710, +1.843455109, +0.000837180, -0.717421988, -0.391289006, "
+      "-0.391289006;+0.166722314, +3.832429713, +0.000847971, +1.913125302, +1.043437349, "
+      "+1.043437349;+0.163968663, +3.335186062, +0.000845273, +0.000000000, +0.000000000, "
+      "+0.000000000;");
```
Comment (MissingNO57) [inline libs/ml/tests/unit/layers/scaled_dot_product_attention.cpp]: Same as previous, I think it's worth of checking.

**Review by MissingNO57 — CHANGES_REQUESTED** [review]
Thank you for fixing my tests. there is one suspicious thing you should look at.

**Review by MissingNO57 — APPROVED** [review]
LGTM
